src/peep/peep.c

```diff
@@ -7772,13 +7772,404 @@ static void peep_on_enter_or_exit_ride(rct_peep *peep, int rideIndex, int flags)
 	}
 }
 
-/**
+/** Main logic to decide whether a peep should buy an item in question
+ *
+ * Seems to also handle the actual purchase as well, so once it returns, the
+ * peep will have the item and the money will have been deducted.
+ *
+ * eax: shopItem | (rideIndex << 8)
+ * ecx: price
+ * esi: peep*
+ *
+ * Returns 0 or 1 depending on if the peep decided to buy the item
  *
  *  rct2: 0x0069AF1E
  */
 static bool sub_69AF1E(rct_peep *peep, int rideIndex, int shopItem, money32 price)
 {
-	return !(RCT2_CALLPROC_X(0x0069AF1E, shopItem | (rideIndex << 8), 0, price, 0, (int)peep, 0, 0) & 0x100);
+#define STANDARD_FOOD_MASK 0x036BA3E0
+#define EXTRA_FOOD_MASK 0x0011FF78
+
+	rct_ride* ride = GET_RIDE(rideIndex);
+
+	///sub_69AF1E
+	RCT2_GLOBAL(0x00F1AEE3, unsigned char) = 0; // Seems to be a local bool for whether the peep has a matching voucher for this shop
+
+	if ((peep->item_standard_flags & PEEP_ITEM_VOUCHER) && // If they have a voucher...
+		(peep->voucher_type == VOUCHER_TYPE_FOOD_OR_DRINK_FREE) && // ...and it's a free food voucher...
+		(peep->voucher_arguments == shopItem)) { // ...and it matches the item in question?...
+		RCT2_GLOBAL(0x00F1AEE3, unsigned char) = 1; // then this becomes true
+	}
+
+///loc_69AF47:
+	// Check to see if the peep already has the item that they're considering buying
+	int shopItem2 = shopItem; // A separate item that's used as output
+	// The root if block you see right below is common in this function.
+	// It seems to be used to branch based on whether this shop serves either a
+	// standard or an extra item. It can probably be replaced with a bool and shopItem
+	// used directly.
+	if (shopItem >= 32) {
+		shopItem2 -= 32;
+		if (peep->item_extra_flags & (1 << shopItem2)) {
+			peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_ALREADY_GOT, shopItem);
+			return 0;
+		}
+	}
+	else {
+		if (peep->item_standard_flags & (1 << shopItem2)) {
+			peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_ALREADY_GOT, shopItem);
+			return 0;
+		}
+	}
+
+///loc_69AF6B:
+	// Checks to see if the peep is already eating or is too nauseous
+	// TODO: check to make sure all of the conditions are covered
+	// TODO: peep_has_food() is redundant with peep_get_standard/extra_item_index()
+	if ((shopItem >= 32 ? EXTRA_FOOD_MASK : STANDARD_FOOD_MASK) & (1 << shopItem2)) { // If the shop offers food
```

Comment:
This line is specifically checking to see if the shop is offering food. AFAIK there is no equivalent set of functions for shops yet.